### ISSUE TITLE
Experimentation with docker

### ISSUE THREAD

**Issue by AlexandreEichenberger** [onnx/onnx-mlir#1078]
Attempting to use the docker tools, and here is some results of my experimentation.

First try, `docker/onnx-mlir.py` tool on a mac

``` shell
(onnx-mlir) alexe@pancetta mnist_v8 % ~/Onnxcode/onnx-mlir/docker/onnx-mlir.py model.onnx
3b46028d19f2: Pulling fs layer
3b46028d19f2: Verifying Checksum
3b46028d19f2: Download complete
3b46028d19f2: Pull complete
cannot open output file 'onnxtempdump86ZNhg': Permission denied
Operation not permitted
```

Same results if I add `--EmitLib`. The `--EmitONNXBasic` works, `--EmitONNXIR`, `--EmitLLVMIR`, `'--EmitJNI` do not, all with a cannot open output file... Tested on LoZ with the same result on Z13 (BITP); on Z14 it failed to pull dockers (probably the machine does not support docker).

Documentation need to state that we can pass additional options? I verified that it works, so we should just add this info.

This script need to be tested in our CIs. If dockers are to be recommended as our preferred approach, we should test all recommended approaches. I know it is a bit awkward, since we publish only once it passes; at the very least getting an error if the about to be published docker is failing, or getting a warning if the published docker is failing this test.

**Comment by AlexandreEichenberger** [comment]
It might be nice also to see if we can build run-on-lib so that users could have a run of the generated .so even if it does not produce any results, or adapt the example of the end-to-end mnist example in the doc to use this facility as well.

**Comment by caoimhinuibrian** [comment]
@AlexandreEichenberger I agree, if this doesn't work it should be fixed. But our discussion about recommending docker for development was not about this python script. It was about pulling the build images from our CI and working inside them.

**Comment by caoimhinuibrian** [comment]
> @AlexandreEichenberger I agree, if this doesn't work it should be fixed. But our discussion about recommending docker for development was not about this python script. It was about pulling the build images from our CI and working inside them.

As opposed to making sure all the prereqs were correctly installed on your machine/OS of choice, cloning the project, running the build scripts ...

**Comment by gongsu832** [comment]
The permission error was due to a bug in `src/Transform/ONNX/ONNXOpTransformPass.cpp`:
```
    101   uint64_t createTagForIR(mlir::ModuleOp module) {
    102     char tempFile[64];
    103 #ifdef _WIN32
    104     strcpy(tempFile, "onnxtempdumpXXXXXX");
    105     _mktemp(tempFile);
    106 #else
    107     strcpy(tempFile, "onnxtempdumpXXXXXX");     <===== this is trying to create under root inside containers
    108     mkstemp(tempFile);
    109 #endif
    110     outputCode(module, tempFile);
    111     uint64_t r = hashFile(tempFile);
    112     remove(tempFile);
    113     return r;
    114   }
```
where it's trying to create a temporary file in the current working directory, which is root inside containers. The proper way is to create a temporary file under /tmp. I have created a PR to fix it.

CI tests are typically designed to test "ground truth", i.e., things that are true regardless of the running environment, like our CI tests for the correct results being produced by the compiled models. Bugs related to the scripts and onnx-mlir binaries themselves are, unfortunately, typically related to specific running environment, which a CI test cannot cover entirely (unless the CI test reproduces all possible running environments).

Also reiterating what Kevin already said, this script is a companion for the user docker image `onnxmlirczar/onnx-mlir`, which obviously is a different issue with the usage of the dev docker image `onnxmlirczar/onnx-mlir-dev` for development.

**Comment by AlexandreEichenberger** [comment]
Thanks, checked that it works
``` shell
(onnx-mlir) alexe@pancetta mnist_v8 % ~/Onnxcode/onnx-mlir/docker/onnx-mlir.py model.onnx
513013d1a882: Pulling fs layer
513013d1a882: Verifying Checksum
513013d1a882: Download complete
513013d1a882: Pull complete
Shared library /Users/alexe/Onnxcode/models/mnist_v8/model.so has been compiled.
```

Will separately update the documentation in a later PR

**Comment by AlexandreEichenberger** [comment]
Actually, I can run the model.so using the run-onnx-lib script with the onnx-mlir generated model.so, but the script does not work when working with the onnx-mlir.py generated model.so.

The failure is that the run-onnx-lib program fail to load the DLL. The code that fails is here:
``` C++
  void *handle = dlopen(name.c_str(), RTLD_LAZY);
```

Below is the objdump -t for the onnx-mlir generated so file, followed by the onnx-mlir.py generated .so file.

```
(onnx-mlir) alexe@pancetta mnist_v8 % objdump -t model.so | grep run_main_graph
00000000000034b0 g     F .text	000001cd run_main_graph
(onnx-mlir) alexe@pancetta mnist_v8 % objdump -t model-direct.so | grep run_main_graph
00000000000031a0 g     F __TEXT,__text _run_main_graph
```

Not sure if the difference between the two entry points is relevant or not.

@gongsu832 were you successful in using the docker generated file?

**Comment by AlexandreEichenberger** [comment]
There is another issue with the docker code, which I am not sure how to fix. 
When using a mnist.onnx, compiling it on docker, and then running it with the python interface, the python fails to get the input_signature functions on the python execution session. The same code works on mac (non-docker version).

Using the mninst.onnx file provided here (which can also be obtained using pytorch instruction in the example)
``` shell
# compile 
onnx-mlir mnist.onnx 
# make sure the PyRuntime has a soft link here from onnx-mlir/build/Debug/lib
ls -l PyRuntime.cpython-38-x86_64-linux-gnu.so 
python mnist.py
# Traceback (most recent call last):
#  File "mnist.py", line 8, in <module>
#     print("input signature in json", session.input_signature())
#  AttributeError: 'PyRuntime.ExecutionSession' object has no attribute 'input_signature'
```

that same example works without issue when not run into the container, on the mac. One difference maybe that it uses gnu? Really not sure.

**Comment by AlexandreEichenberger** [comment]
[mnist.onnx.gz](https://github.com/onnx/onnx-mlir/files/7825117/mnist.onnx.gz)

**Comment by gongsu832** [comment]
@AlexandreEichenberger .so generated by the docker container is a Linux ELF, which I don't expect to be usable on MacOS.

**Comment by AlexandreEichenberger** [comment]
@gongsu832 same for windows? The `onnx-mlir.py model.onnx` generates a Linux ELF binary? 
Its a bit confusing because the expectation of running a `onnx-mlir.py` is that it would mimic `onnx-mlir`. But with a warning, it can be made clear...

**Comment by gongsu832** [comment]
I don't know much about Windows but yes I would expect the same, i.e., the .so is not directly usable on the host. Yes we can add a warning if the host is not Linux.

**Comment by AlexandreEichenberger** [comment]
thanks for your input
